### PR TITLE
Implement support for Projects and defaults for user, password and codeset

### PR DESCRIPTION
**Pull request by jsuchome**
When CurrentProject is set in .fuseml/config.yaml, commands normally requiring "--project" option will use CurrentProject when "--project" is not provided.

implementation of https://github.com/fuseml/fuseml/issues/118

Example of `.fuseml/config.yaml` (note that it's not case sensitive):

```
currentcodeset: mlflow-02
currentproject: mlflow-project-02
url: http://localhost:8000
currentuser: fuseml-mlflow-project-02
currentpassword: newpassword
```

It's just one level for now, we can make it more structured if we find it useful.

The priorities go like: CLI option > env variable > config file under current directory > config file under $HOME

### CODE REVIEW

File: pkg/cli/cmd.go [hunk 1]
```diff
@@ -95,6 +95,10 @@ func bindFlags(cmd *cobra.Command) {
 			val := viper.Get(f.Name)
 			cmd.Flags().Set(f.Name, fmt.Sprintf("%v", val))
 		}
+		// use CurrentProject as a backup for missing project flag
+		if !f.Changed && (f.Name == "project" || f.Name == "codeset-project") && viper.IsSet("CurrentProject") {
+			cmd.Flags().Set(f.Name, viper.GetString("CurrentProject"))
+		}
```
Comment (jsuchome) [inline pkg/cli/cmd.go:101]: The problem of doing it here is that I'll need to check which command is being executed, for example it's not possible to fill in "name" argument value, as it could be name of different artefact...

File: design/project.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+	Field(3, "description", String, "Project description", func() {
+		Example("Set of MLFlow applications")
+		Default("")
+	})
+	Required("name")
+})
+
+// User describes the user assigned to the project
+var User = Type("User", func() {
+	Field(1, "name", String, "User name", func() {
```
Comment (flaviodsr) [inline design/project.go:121]: Is this the name of the user (Full Name) or the username?

File: pkg/cli/project/delete.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+// NewSubCmdProjectDelete creates and returns the cobra command for the `project delete` CLI command
+func NewSubCmdProjectDelete(gOpt *common.GlobalOptions) *cobra.Command {
+
+	o := NewDeleteOptions(gOpt)
+
+	cmd := &cobra.Command{
+		Use:   `delete {-n|--name NAME}`,
+		Short: "Delete projects.",
+		Long:  `Delete a project from FuseML`,
```
Comment (flaviodsr) [inline pkg/cli/project/delete.go:33]: I think it would be nice to say that deleting a project also deletes all codesets associated with the project, or it cannot be deleted if there are existing codesets associated with it?

File: pkg/cli/project/cmd.go [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+import (
+	"github.com/fuseml/fuseml-core/pkg/cli/common"
+	"github.com/spf13/cobra"
+)
+
+// NewCmdProject creates and returns the cobra command that acts as a root for all other project CLI sub-commands
+func NewCmdProject(c *common.GlobalOptions) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "project",
+		Short: "project management",
```
Comment (flaviodsr) [inline pkg/cli/project/cmd.go:12]: Probably should start with uppercase `Project`.

File: pkg/cli/project/list.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+	return
+}
+
+// NewSubCmdProjectList creates and returns the cobra command for the `project list` CLI command
+func NewSubCmdProjectList(gOpt *common.GlobalOptions) *cobra.Command {
+
+	o := NewListOptions(gOpt)
+
+	cmd := &cobra.Command{
+		Use:   "list",
```
Comment (flaviodsr) [inline pkg/cli/project/list.go:56]: missing optional name argument.

File: pkg/cli/project/set-current.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+		return nil
+	}
+
+	viper.Set("CurrentProject", o.Name)
+
+	if err := viper.WriteConfig(); err != nil {
+		return err
+	}
+
+	fmt.Printf("Project '%s' saved as current.\n", o.Name)
```
Comment (flaviodsr) [inline pkg/cli/project/set-current.go:64]: This wording sounds a bit weird to me, as a suggestion I would rephrase this something like `Current project set to %s`.

Side note: IMO, using `default` (as is on the PR title) instead of `current` would fit better, but I do not want to ask you to rewrite everything using `default` instead of `current`.

**Review by flaviodsr — CHANGES_REQUESTED** [review]

File: design/project.go [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+			Field(1, "name", String, "List only Project with given name", func() {
+				Example("mlflow-project-01")
+			})
+		})
+
+		Result(ArrayOf(Project), "Return all registered Projects.")
+
+		Error("NotFound", func() {
+			Description("If the Project is not found, should return 404 Not Found.")
+		})
```
[the remark this answers is not in the document]
Comment (jsuchome) [reply design/project.go:25]: You are right

File: design/project.go [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+			})
+			Required("name")
+		})
+
+		Error("BadRequest", func() {
+			Description("If name is not given, should return 400 Bad Request.")
+		})
+		Error("NotFound", func() {
+			Description("If there is no project with the given name, should return 404 Not Found.")
+		})
```
[the remark this answers is not in the document]
Comment (jsuchome) [reply design/project.go:91]: Hm, I guess you are right

File: design/project.go (hunk 2 above)
Comment (jsuchome) [reply design/project.go:121]: User name

File: pkg/cli/project/delete.go (hunk 3 above)
Comment (jsuchome) [reply pkg/cli/project/delete.go:33]: Actually, I just remembered that I did not implement delete operation yet.... 

But I think i'd rather not allow deleting project when codesets still exist

File: pkg/cli/project/set-current.go (hunk 6 above)
Comment (jsuchome) [reply pkg/cli/project/set-current.go:64]: You are right about wording. 

I'm not sure about the other part

File: pkg/cli/codeset/set-current.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+	return res
+}
+
+// NewSubCmdCodesetSetCurrent creates and returns the cobra command for the `codeset set` CLI command
+func NewSubCmdCodesetSetCurrent(gOpt *common.GlobalOptions) *cobra.Command {
+
+	o := NewSetCurrentOptions(gOpt)
+
+	cmd := &cobra.Command{
+		Use:   `set-current {-n|--name NAME}`,
```
Comment (stefannica) [inline pkg/cli/codeset/set-current.go:31]: Same here as with `project set`: why not use `codeset set` instead ?

File: pkg/cli/project/list.go [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+	if project, ok := object.(*project.Project); ok {
+		for _, user := range project.Users {
+			if formated != "" {
+				formated += "\n"
+			}
+			formated += fmt.Sprintf("- name: %s\n  email: %s", user.Name, user.Email)
+		}
+	}
+	return
+}
```
Comment (stefannica) [inline pkg/cli/project/list.go:36]: Thanks for implementing a custom formatter !

File: pkg/cli/codeset/register.go [hunk 11]
```diff
@@ -53,6 +56,13 @@ LOCATION can be path to local directory or URL of a git repository`,
 	cmd.Flags().StringVarP(&o.Project, "project", "p", "", "the project to which the codeset belongs")
 	cmd.Flags().StringVarP(&o.Description, "desc", "d", "", "codeset description")
 	cmd.Flags().StringSliceVar(&o.Labels, "label", []string{}, "one or more codeset labels associated with the codeset")
+
+	cmd.Flags().StringVarP(&o.Password, "password", "", "", "(FUSEML_PROJECT_PASSWORD) Password of the user accessing a project")
```
Comment (stefannica) [inline pkg/cli/codeset/register.go:60]: Here's an idea: shouldn't we automatically set the project and codeset supplied here as the current ones ? I don't fully understand the user and password bits, but I figure that also applies to them.

File: pkg/cli/cmd.go [hunk 12]
```diff
@@ -95,6 +126,20 @@ func bindFlags(cmd *cobra.Command) {
 			val := viper.Get(f.Name)
 			cmd.Flags().Set(f.Name, fmt.Sprintf("%v", val))
 		}
+		// use CurrentProject as a backup for missing project flag
+		if !f.Changed && defaultForMissingFlagPossible(f.Name, "CurrentProject") {
```
Comment (stefannica) [inline pkg/cli/cmd.go:130]: I wanted to try it, but didn't get the chance yet: what happens if you don't have a configuration file when you run `project set` ? Is one created automatically ?

File: pkg/cli/common/global.go [hunk 13]
```diff
@@ -19,6 +19,10 @@ type GlobalOptions struct {
 	Timeout int
 	// Verbose mode prints out additional information
 	Verbose bool
+	// CurrentProject says which project to use if "project" flag is not passed
+	CurrentProject string
+	// CurrentCodeset sets which codeset to use if the name is not provided
+	CurrentCodeset string
```
Comment (stefannica) [inline pkg/cli/common/global.go:25]: This is related to the crazy idea I proposed in an earlier comment: try naming these `Project` and `Codeset`, then linking them to the individual flags where they're needed in the sub-commands. This way, these variables don't only play the role of "current" project and codeset, but they also hold the value if explicitly supplied as a command like argument.

File: pkg/cli/cmd.go [hunk 14]
```diff
@@ -84,6 +86,35 @@ func initializeConfig(cmd *cobra.Command) error {
 	return nil
 }
 
+func defaultForMissingFlagPossible(flag, defaultKey string) bool {
```
Comment (stefannica) [inline pkg/cli/cmd.go:89]: I see you chose to implement a global mechanism for using default project and codeset values if they're not explicitly supplied as arguments where they're needed for their respective sub-commands. This is one way to do it, but now you have two places where you need to keep flags in sync for applicable sub-commands: here and where the cobra command is defined.

I would have spread it out across all commands where this is applicable instead of in a central location. That way, it's more obvious where you need this particular type of behavior. I'm not saying you should duplicate this code everywhere you have a codeset or project argument, but it should be something that you can configure for each cobra command explicitly, preferably in the same area of code where you define the flags.

Here's a crazy idea: what if you're "linking" all those individual flags to the same global variables ? For example, instead of linking the `workflow assign {-p|--codeset-project}` flag to the local `assignOptions.codesetProject` variable, link it to the global one that is also linked to the currently set project. That way, you only have one variable you need to care about for all flags that point to a project _and_ you get some indication in each cobra command that this is the behavior you want. It might take some hacking to make it work with the concept of a mandatory flag, but I think that's doable.

**Review by stefannica — CHANGES_REQUESTED** [review]
Great work ! Many thanks especially for putting in the extra time to implement a standalone endpoint and service for projects on the server side. The only thing needed to make it into a complete project management component is "project create" and I highly recommend that you include that in your PR too, to not break the separation of concerns principle.

File: design/project.go [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+
+import (
+	. "goa.design/goa/v3/dsl"
+)
+
+var _ = Service("project", func() {
+	Description("The project service performs operations on Projects.")
+
+	// Method describes a service method (endpoint)
+	Method("list", func() {
```
[the remark this answers is not in the document]
Comment (jsuchome) [reply design/project.go:11]: OK, good point actually

File: pkg/cli/project/set-current.go [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+	return res
+}
+
+// NewSubCmdProjectSetCurrent creates and returns the cobra command for the `project set` CLI command
+func NewSubCmdProjectSetCurrent(gOpt *common.GlobalOptions) *cobra.Command {
+
+	o := NewSetCurrentOptions(gOpt)
+
+	cmd := &cobra.Command{
+		Use:   `set-current {-n|--name NAME}`,
```
[the remark this answers is not in the document]
Comment (jsuchome) [reply pkg/cli/project/set-current.go:31]: well, I had `set` originally, but then it looked like `set-current` fits better to what is actually done...

File: pkg/core/project_store.go [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+import (
+	"context"
+
+	"github.com/pkg/errors"
+
+	"github.com/fuseml/fuseml-core/pkg/domain"
+)
+
+// GitProjectStore describes a stucture that accesses project store implemented in git
+type GitProjectStore struct {
```
[the remark this answers is not in the document]
Comment (jsuchome) [reply pkg/core/project_store.go:12]: I do not think it as so different than other artifacts that it would require different treatment...

Manager might be better but maybe I'll just refactor it again after we have manager implemented for workflows

File: pkg/cli/codeset/register.go (hunk 11 above)
Comment (jsuchome) [reply pkg/cli/codeset/register.go:60]: Hm... I'm not sure. Maybe it could be nice.
But it also could be bit confusing, doing something user did not ask for...

File: pkg/cli/cmd.go (hunk 12 above)
Comment (jsuchome) [reply pkg/cli/cmd.go:130]: Good question, I did not test such particular case. I would assume it gets created, since it uses standard `viper` library calls and they should take care of it.

File: pkg/cli/cmd.go (hunk 14 above)
Comment (jsuchome) [reply pkg/cli/cmd.go:89]: OK, maybe I'm missing something but this seems like I should treat Project/Codeset as global options, like e.g. `url` option and related `FUSEML_SERVER_URL` env var.

Then it would work without that ugly `defaultForMissingFlagPossible` processing, however, it would reveal global Project/Codeset options to all commands, even those where they are not needed.

If, on the other hand, I would completely ignore it on the global level and only try to handle it on the specific subcommand level, I would have a problem (as you hinted) with those flags marked as mandatory because the check for mandatory flags is processed before parsing the config file ....

_but it should be something that you can configure for each cobra command explicitly, preferably in the same area of code where you define the flags._

Even if the problem with mandatory flags could be avoided, I'm not sure about this part. This would require extra config file processing for each subcommand, apart from the one that is done globaly

Comment (jsuchome) [reply pkg/cli/cmd.go:89]: Let me phrase it yet differently: I'm quite sure it could be done with other approach than the one I chose and that it could be better in some aspects. 
However I have a feeling it would mean just some different 'hackish' (read: ugly) solutions in different parts of the code then the ones I modified.

I could be wrong of course, it took me some time to understand how this cobra/viper combination works and I could have missed something obvious.

File: pkg/cli/client/common.go [hunk 18]
```diff
@@ -50,11 +52,12 @@ func (c *Clients) InitializeClients(URL string, timeout int, verbose bool) error
 		doer = goahttp.NewDebugDoer(doer)
 	}
 
-	c.CodesetClient = codesetc.NewClient(scheme, host, doer, encoder, decoder, verbose)
 	c.ApplicationClient = applicationc.NewClient(scheme, host, doer, encoder, decoder, verbose)
-	c.WorkflowClient = NewWorkflowClient(scheme, host, doer, encoder, decoder, verbose)
+	c.CodesetClient = codesetc.NewClient(scheme, host, doer, encoder, decoder, verbose)
+	c.ProjectClient = projectc.NewClient(scheme, host, doer, encoder, decoder, verbose)
```
[the remark this answers is not in the document]
Comment (jsuchome) [reply pkg/cli/client/common.go:57]: I actually thought about it, but then decided to go the traditional way. I may still address this.

File: pkg/cli/codeset/register.go [hunk 19]
```diff
@@ -53,6 +56,13 @@ LOCATION can be path to local directory or URL of a git repository`,
 	cmd.Flags().StringVarP(&o.Project, "project", "p", "", "the project to which the codeset belongs")
```
Comment (stefannica) [inline pkg/cli/codeset/register.go:56]: Folow-up to my suggestion about using global variables: what would happen if you did this here instead ?

```
cmd.Flags().StringVarP(&gOpt.CurrentProject, "project", "p", "", "the project to which the codeset belongs")
```

that's what I meant by linking the flag to the global option variable. This way, you can configure this on a per-command basis instead of aggregating everything in that `cli/cmd.go` common file, transparently from each command. Generally speaking, transparent mechanisms like this are more difficult to debug and maintain.

There's another reason for it: the way that I see it, this CLI code is built on top of something resembling a framework that people can easily use to write new commands. A "current project" is a global CLI feature that can apply to any command, so you should implement something to make it easy for anyone to use it when writing new commands. Something as simple as a function call that says "link this flag to the current project".

Comment (jsuchome) [reply pkg/cli/codeset/register.go:56]: value of `gOpt.CurrentProject` is not yet known at this time, because it's only filled when parsing the command line (`bindFlags`)

File: pkg/cli/client/common.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (jsuchome) [reply pkg/cli/client/common.go:57]: See next commit. Reused @flaviodsr code to the bits, thanks for inspiration!

File: pkg/cli/project/set-current.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (stefannica) [reply pkg/cli/project/set-current.go]: thanks :smile: , I'm forever grateful for it

File: pkg/cli/codeset/register.go (hunk 19 above)
Comment (stefannica) [reply pkg/cli/codeset/register.go:56]: ok, let's leave this for some other time, good enough for now

**Review by stefannica — APPROVED** [review]

**Review by stefannica — DISMISSED** [review]

**Review by stefannica — APPROVED** [review]

**Review by flaviodsr — APPROVED** [review]
LGTM
